sdk/ai/Azure.AI.Inference/src/Telemetry/OpenTelemetryScope.cs

```diff
@@ -0,0 +1,247 @@
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="name">The name of tag to set.</param>
+        /// <param name="value">Nullable value to be set.</param>
+        private void SetTagMaybe<T>(string name, Nullable<T> value) where T: struct
+        {
+            if (value.HasValue && m_activity != null)
+                m_activity.SetTag(name, value.Value);
+        }
+
+        /// <summary>
+        /// Return true if the activity for OpenTelemetry exists.
+        /// </summary>
+        public bool IsEnabled { get => m_activity != null; }
+
+        /// <summary>
+        /// Return true if the metrics are enabled.
+        /// </summary>
+        public bool AreMetricsOn { get => s_duration.Enabled && s_tokens.Enabled; }
+
+        /// <summary>
+        /// Create the instance of OpenTelemetryScope. This constructor logs request
+        /// and starts the execution timer.
+        /// </summary>
+        /// <param name="requestOptions">The request options used in the call.</param>
+        /// <param name="endpoint">The endpoint being called.</param>
+        public OpenTelemetryScope(ChatCompletionsOptions requestOptions, Uri endpoint)
+        {
+            m_recordedStreamingResponse = new();
+            m_activity = GetActivityMaybe($"Complete_{requestOptions.Model}");
+            if (m_activity == null)
+                return;
+            m_activity.Start();
+            // Record the request to telemetry;
+            m_commonTags = new TagList{
+                { GenAiSystemKey, GenAiSystemValue},
+                { GenAiResponseModelKey, requestOptions.Model},
+                { ServerAddressKey, endpoint.Host },
+                { ServerPortKey, endpoint.Port },
+                { GenAiOperationNameKey, "Complete"}
+            };
+            // Set tags for reporting them to console.
+            foreach (KeyValuePair<string, object> kv in m_commonTags)
+            {
+                m_activity.SetTag(kv.Key, kv.Value);
+            }
+            SetTagMaybe(GenAiRequestMaxTokensKey, requestOptions.MaxTokens);
+            SetTagMaybe(GenAiRequestTemperatureKey, requestOptions.Temperature);
+            if (requestOptions.AdditionalProperties != null && requestOptions.AdditionalProperties.ContainsKey("top_p"))
+            {
+                SetTagMaybe(
+                    GenAiRequestTopPKey,
+                    JsonSerializer.Deserialize<double?>(requestOptions.AdditionalProperties["top_p"]));
+            }
+            // Log all messages as the events.
+            foreach (ChatRequestMessage message in requestOptions.Messages)
+            {
+                TagList requestTags = new() {
+                    { GenAiSystemKey, GenAiSystemValue},
+                    { GenAiEventContent, GetContent(message)}
+                };
+                m_activity.AddEvent(
+                    new ActivityEvent(
+                        $"gen_ai.{message.Role}.message",
```

Comment:
is `message.Role` a string? how does it look like? (i.e. `user` or `User`?)

Reply:
It will we `user` or `system` from then lower case.